### PR TITLE
Choose HDFS read API based on Client buffer size

### PR DESCRIPTION
**Pull request by madanadit**
This PR is a follow up to https://github.com/Alluxio/alluxio/pull/10763.

When the client buffer size is large ( > 2MB) and reads are guaranteed to be somewhat sequential, the `pread` API to HDFS is not as efficient as simple `read`. We introduce a heuristic to choose which API to use.

### CODE REVIEW

**Comment by madanadit** [comment]
@calvinjia Early comments requested. The performance seems to workout for large buffer sizes with this PR. 

A couple of open questions:
1. Should the threshold be made a mount option? The threshold should depend on how remote the HDFS is, for slower HDFS I could imagine a lower threshold could work better. The disadvantage of this over the current client option is that we cannot change this on the fly (with path defaults)
2. If using a mount option, instead of comparing against the threshold on the client side, I would then pass in `pread_buffer_size` as part of `ReadRequest`, propagate to the UFS  and  evaluate the condition in the worker ufs

Let me know your thoughts.

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java [hunk 1]
```diff
@@ -233,6 +234,11 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
       return -1;
     }
 
+    LOG.info("AMDEBUG pread len {} ", len);
+    if (len < 2 * Constants.MB) {
+      LOG.info("AMDEBUG set for pread len {} ", len);
+      mOptions.setPositionShort(true);
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java]: Does this option only apply to positioned reads or also reads through the `read` API?

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java [hunk 2]
```diff
@@ -233,6 +234,11 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
       return -1;
     }
 
+    LOG.info("AMDEBUG pread len {} ", len);
+    if (len < 2 * Constants.MB) {
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java]: I think it is good to have our heuristic to cover the typical case, but should we check if it has been set to allow the user to override?

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java [hunk 3]
```diff
@@ -563,10 +563,21 @@ public InputStream open(String path, OpenOptions options) throws IOException {
...
-        // pread API instead of seek is more efficient for FSDataInputStream.
-        // A seek on FSDataInputStream uses a skip op which is implemented as read + discard
-        // and hence ends up reading extra data from the datanode.
-        return new HdfsPositionedUnderFileInputStream(inputStream, options.getOffset());
+        if (options.getPositionShort()) {
+          LOG.info("AMDEBUG: using pread");
+          // pread API instead of seek is more efficient for FSDataInputStream.
+          // A seek on FSDataInputStream uses a skip op which is implemented as read + discard
+          // and hence ends up reading extra data from the datanode.
+          return new HdfsPositionedUnderFileInputStream(inputStream, options.getOffset());
```
Comment (calvinjia) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: Do we end up caching this stream anywhere, if so would it be a problem if it is a certain type?

**Comment by calvinjia** [comment]
@madanadit I think a user side option is more flexible and we can have a server side default instead of the client side default if we want it to be more user friendly.

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java (hunk 1 above)
Comment (madanadit) [reply core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java]: only applies to `pread`. reads through the `read` API always are assumed to be sequential by this heuristic

File: core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java (hunk 2 above)
Comment (madanadit) [reply core/client/fs/src/main/java/alluxio/client/file/AlluxioFileInStream.java]: yep, i'll make this a client config which can be overridden by the user

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java (hunk 3 above)
Comment (madanadit) [reply underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: I just grepped our codebase, doesn't look like we do, so we should be fine

File: tests/src/test/java/alluxio/client/fs/FileInStreamIntegrationTest.java [hunk 4]
```diff
@@ -435,6 +435,44 @@ public void positionedReadWithoutCaching() throws Exception {
...
+      String filename = mTestPath + "/file_" + MIN_LEN + "_" + op.hashCode();
+      AlluxioURI uri = new AlluxioURI(filename);
+
+      try (FileInStream is = mFileSystem.openFile(uri, FileSystemTestUtils.toOpenFileOptions(op))) {
+        byte[] ret = new byte[DELTA - 1];
+        Assert.assertEquals(DELTA - 1, is.positionedRead(MIN_LEN - DELTA + 1, ret, 0, DELTA));
+        Assert.assertTrue(BufferUtils.equalIncreasingByteArray(MIN_LEN - DELTA + 1, DELTA - 1, ret));
+      }
+    }
+  }
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/client/fs/FileInStreamIntegrationTest.java]: Is there a way to verify if the PRead path is utilized?

**Comment by madanadit** [comment]
jenkins, test this please.

**Comment by madanadit** [comment]
alluxio-bot, merge this please.

**Pull request merged**